data/api_v1_data_helper.go

```diff
@@ -0,0 +1,230 @@
+	timeIt(ctx, "getParamHistory")
+	writer.parametersHistory, err = a.store.GetDiabeloopParametersHistory(ctx, userID, parameterLevelFilter[:])
+	if err != nil {
+		// Just log the problem, don't crash the query
+		writer.parametersHistory = nil
+		a.logger.Printf("{%s} - {GetDiabeloopParametersHistory:\"%s\"}", traceID, err)
+	}
+	timeEnd(ctx, "getParamHistory")
+	return iterPumpSettings, nil
+}
+
+func (a *API) writeDataV1(
+	ctx context.Context,
+	res *httpResponseWriter,
+	includePumpSettings bool,
+	iterPumpSettings mongo.StorageIterator,
+	iterUploads mongo.StorageIterator,
+	iterData mongo.StorageIterator,
+	tideV2Data []schema.CbgBucket,
+	writeParams *writeFromIter,
+) error {
+	timeIt(ctx, "writeData")
+	defer timeEnd(ctx, "writeData")
+	// We return a JSON array, first charater is: '['
+	err := res.WriteString("[\n")
+	if err != nil {
+		return err
+	}
+
+	if includePumpSettings && iterPumpSettings != nil {
+		writeParams.iter = iterPumpSettings
+		err = writeFromIterV1(ctx, writeParams)
+		if err != nil {
+			return err
+		}
+	}
+
+	timeIt(ctx, "writeDataMain")
+	writeParams.iter = iterData
+	err = writeFromIterV1(ctx, writeParams)
+	if err != nil {
+		return err
+	}
+	timeEnd(ctx, "writeDataMain")
+
+	if len(tideV2Data) > 0 {
+		timeIt(ctx, "writeDataV2")
+		writeParams.dataV2 = tideV2Data
+		err = writeDataV2(ctx, writeParams)
+		if err != nil {
+			return err
+		}
+		timeEnd(ctx, "writeDataV2")
+	}
+	// Fetch uploads
+	if len(writeParams.uploadIDs) > 0 {
+		timeIt(ctx, "getUploads")
+		iterUploads, err = a.store.GetDataFromIDV1(ctx, res.TraceID, writeParams.uploadIDs)
+		if err != nil {
+			// Just log the problem, don't crash the query
+			writeParams.parametersHistory = nil
+			a.logger.Printf("{%s} - {GetDataFromIDV1:\"%s\"}", res.TraceID, err)
+		} else {
+			defer iterUploads.Close(ctx)
+			writeParams.iter = iterUploads
+			err = writeFromIterV1(ctx, writeParams)
+			if err != nil {
+				timeEnd(ctx, "getUploads")
+				return err
+			}
+		}
+		timeEnd(ctx, "getUploads")
+	}
+
+	// Silently failed theses error to the client, but record them to the log
+	if writeParams.decode.firstError != nil {
+		a.logger.Printf("{%s} - {nErrors:%d,MongoDecode:\"%s\"}", res.TraceID, writeParams.decode.numErrors, writeParams.decode.firstError)
+	}
+	if writeParams.jsonError.firstError != nil {
+		a.logger.Printf("{%s} - {nErrors:%d,jsonMarshall:\"%s\"}", res.TraceID, writeParams.jsonError.numErrors, writeParams.jsonError.firstError)
+	}
+
+	// Last JSON array charater:
+	return res.WriteString("]\n")
+}
+
+// Mapping V2 Bucket schema to expected V1 schema + write to output
+func writeDataV2(ctx context.Context, p *writeFromIter) error {
+	var err error
+	for _, bucket := range p.dataV2 {
+		for i, sample := range bucket.Samples {
+			datum := make(map[string]interface{})
+			// Building a fake id (bucket.Id/range index)
+			datum["id"] = fmt.Sprintf("%s_%d", bucket.Id, i)
+			datum["type"] = "cbg"
+			datum["uploadId"] = bucket.Id
```

Comment:
based on a discussion with Yvon, this may not be necessary (check with him but he told me that this value is re-computed in the front if missing ...)